### ISSUE TITLE
feat(reviewing-code): Marasa-Trois — empirical falsification of LLM dissent (Phase 2.6, complements #807 + #810)

### ISSUE THREAD

**Issue by deep-name** [0xHoneyJar/loa#811]
### Summary

The current `reviewing-code` Phase 2.5 (`adversarial-review.sh`) collects findings from N LLM dissenters and treats their consensus (or single-model verdict) as binding. This works when the dissenters' errors are **independent** — Condorcet's jury theorem guarantees consensus accuracy approaches 1 as N grows, *given* better-than-random + independent evaluators.

**LLMs trained on similar substrates are not independent.** Two models can be confidently-wrong in correlated ways from shared training-data myths. The framework's existing `hallucination_filter` is statistical and does not catch plausible-sounding-but-empirically-wrong technical claims.

This proposal adds a third voice to the dialectic: a non-LLM oracle (bash, the typechecker, the protocol implementation) that can falsify LLM dissent claims by *behaving*, not arguing. The pattern is named **Marasa-Trois** — the invisible third twin in Haitian Vodou, who completes what the two visible twins cannot. The mnemonic encodes the invariant: when an agent sees the name, they remember to *run the test, not just collect the dissent*.

### The empirical incident — sprint-0b of `loa-vps-setup`

In sprint-0b round 3 of `0xHoneyJar/loa-vps-setup` PR #2 (a CLI ergonomics + security hardening sprint), `claude-sonnet-4-6` returned 4 BLOCKING findings about bash semantics. The reviewer manually falsified 3 via shell tests:

- **DISS-001/002**: claimed `local rc=$?` always returns 0 because `local` is a command. Empirically: `bash -c 'f(){ (exit 5); local rc=$?; [ $rc = 5 ]; }; f' && echo "rc was 5"` succeeds. Bash expands `$?` at parse time *before* running `local`. Held-lock test confirmed `rc=5` correctly captured.
- **DISS-003**: claimed `|| true` only catches `jq`'s exit code, not `grep`'s, so all-blank-lines coord-state.jsonl would abort the script under `set -euo pipefail`. Empirically: `|| true` on the outer `$()` substitution catches the entire pipeline's exit code (pipefail-amplified grep-exit-1 included). Verified by `printf '\n\n\n' > test.jsonl; result=$(grep -v '^[[:space:]]*$' test.jsonl 2>/dev/null | jq -s '.' 2>/dev/null || true); echo "result=$result; after"` printing `result=[]; after` cleanly.

Without empirical refutation, these findings would have driven the engineer to "fix" working code — introducing real bugs to satisfy phantom ones. The trajectory log shows `hallucination_filter.applied: true, downgraded: 0`. Statistical filters are the wrong tool for this class of error.

### Why this isn't covered by #807 or #810

- **#807** ("Multi-model adversarial review fallback approves issues single-model misses") addresses the *single-dissenter fallback* failure mode. Marasa-Trois is a deeper fix: even with ≥2 dissenters present, they can be correlated-wrong. Empirical refutation is orthogonal to dissenter count.
- **#810** ("consensus voting can suppress single-model security true-positives") addresses *consensus suppressing true-positives* (one model right, others wrong → finding gets dropped). Marasa-Trois is the **dual**: consensus surfacing false-positives (multiple models confidently-wrong → finding gets binding). Both #810 and this proposal converge on the same root cause: **LLM consensus is not a sufficient ground-truth oracle**. Empirical substrate is the missing third voice.

The three issues compose:
- **#807**: ensure ≥2 dissenters actually run
- **#810**: don't drop a single-model finding just because others disagree
- **this**: don't bind a multi-model finding without empirical falsification when the claim is testable

### The pattern

```
┌─────────────────────────────────────────────────────────────┐
│  Phase 2.5  —  LLM Dissent (existing)                       │
│                                                             │
│  Voice A: claude-sonnet-4-6  →  4 BLOCKING findings         │
│  Voice B: gpt-5.3-codex      →  0 findings                  │
│                                                             │
│  Consensus mechanism: governance-isomorphism +              │
│                       deliberative-council                  │
└─────────────────────────────────────────────────────────────┘
                            ↓
┌─────────────────────────────────────────────────────────────┐
│  Phase 2.6  —  Marasa-Trois (proposed)                      │
│                                                             │
│  For each BLOCKING finding with a testable claim:           │
│    1. Derive a bash predicate (LLM or operator)             │
│    2. Execute predicate in sandbox (timeout + tempdir CWD)  │
│    3. Classify: FALSIFIED / VERIFIED / INCONCLUSIVE         │
│    4. Log to marasa-trials-<date>.jsonl                     │
│                                                             │
│  Verdict: finding becomes binding only if !FALSIFIED        │
│           INCONCLUSIVE → operator judgment                  │
└─────────────────────────────────────────────────────────────┘
```

### Working reference implementation

A working tool exists downstream in `0xHoneyJar/loa-vps-setup` PR #3:

- **Tool**: [`bin/marasa-trois.sh`](https://github.com/0xHoneyJar/loa-vps-setup/blob/marasa-trois-bootstrap/bin/marasa-trois.sh) — ~280 lines bash, executable. Sandboxed via `timeout` + tempdir CWD. Audit log uses fd-based-flock (cross-pollinated from CRIT-4/5 fix in PR #2). Verdicts: `FALSIFIED` / `VERIFIED` / `INCONCLUSIVE`.

- **Live demo** — falsifying the actual round-3 hallucinations:

  ```
  $ bin/marasa-trois.sh \
      --finding-id DISS-001 \
      --claim 'local rc=$? always returns 0 because local clobbers $?' \
      --predicate 'f(){ (exit 5); local rc=$?; [ $rc = 5 ]; }; f' \
      --expected-falsified

  ═══ Marasa-Trois — Verdict ═══
    Verdict:    FALSIFIED
    Reason:     predicate succeeded; LLM claim contradicted by empirical substrate
    Exit code:  0
    Duration:   3ms
  ```

- **Audit log** at `grimoires/loa/a2a/trajectory/marasa-trials-<date>.jsonl` accumulates a corpus of falsified-by-empirical-test LLM claims. Two records already populated from running against round-3 findings.

- **Epistemic anchor tests** at `tests/cli-snapshots/run-all.sh` pin the empirical truths (`local rc=$?`, `|| true` pipefail) so future LLM dissent re-raising the same hallucination has an in-repo refutation to point at.

The downstream PR is intentionally narrow: one tool + two anchor tests + lore/vision documentation. No integration into upstream skills yet — that's a follow-up cycle.

### Suggested upstream integration

If the proposal is accepted, an integration cycle could:

1. **Land `bin/marasa-trois.sh` (or equivalent)** as a framework-level tool, possibly under `.claude/scripts/marasa-trois.sh` to fit the existing scripts layout.
2. **Update `.claude/skills/reviewing-code/SKILL.md` Phase 2.5** to add a Phase 2.6 step: for each BLOCKING finding, attempt to derive an empirical predicate. The predicate can come from:
   - LLM proposal (with operator-readable in <5 seconds — if longer, surface for review)
   - Operator-supplied (when the LLM can't derive one)
3. **Schema**: extend the adversarial-review JSONL with a `marasa_trois_verdict` field per finding. Findings with `verdict: FALSIFIED` are logged and dropped; `VERIFIED` proceed through normal flow; `INCONCLUSIVE` fall back to operator judgment (the existing default).
4. **Composition with `red-team-code-vs-design.sh`**: the red-team gate is already a kind of empirical-substrate gate (running the diff against the SDD). Marasa-Trois is the per-finding analogue. The two compose well — red-team for design conformance, Marasa-Trois for individual claim verification.
5. **Falsified-claims corpus**: aggregating `marasa-trials-*.jsonl` across cycles produces a dataset of "claims LLMs made that were empirically wrong." Useful for evaluating future dissenter models, fine-tuning hallucination filters, and characterizing per-model failure modes.

### Honest scope discussion

**What Marasa-Trois cannot do:**

- Decide claims about future maintainability, code style, or other non-testable assertions. Returns INCONCLUSIVE → operator judgment. This is the right semantics — predicates exist for the testable subset only.
- Defend against malicious predicates. The sandbox (`timeout` + tempdir CWD) is defense against *accidental* state mutation by hallucinated predicates, not against adversarial code execution. Treat the predicate the same way you treat a `git apply`: read it before running it.
- Replace operator judgment. The pattern adds a voice; it does not claim final authority.

**Sub-problem: predicate derivation.** The LLM-derives-predicate step is itself fallible. Mitigation in the reference impl: require the predicate be a single bash command the operator can read in <5 seconds; longer or multi-step predicates surface for operator review before execution. This trades "LLM hallucinates a finding" for "LLM hallucinates a predicate, but operator can audit the predicate in seconds." A deeper mitigation would be to derive predicates from a constrained DSL or from the codebase's existing test infrastructure (call existing test functions rather than freeform shell), but the constrained-shell-string variant is the simplest first cut.

**Non-deterministic claims** (race conditions, network-dependent findings, etc.) are the genuinely hard case. Current behavior: predicate timeout → INCONCLUSIVE. Future improvement: parameterized retries (run N times, classify as `MOSTLY_FALSIFIED` / `MOSTLY_VERIFIED` if results converge, `NON_DETERMINISTIC` if they don't). Not in scope for v1.

### Local provenance

Full project-local lineage of this pattern (in `loa-vps-setup`):

- Vision: `grimoires/loa/visions/entries/vision-019-marasa-trois-empirical-dissent-falsification.md`
- Lore: `grimoires/loa/lore/patterns.yaml#marasa-trois-empirical-dissent` (status `Captured`, `upstream_candidate: true`)
- Memory: `grimoires/loa/memory/observations.jsonl` first entry
- Letter to next agent: `grimoires/loa/a2a/sprint-0b/letter-to-the-next-agent.md`

These are gitignored (operator-local artifacts) but the schema mirrors the upstream `.loa/grimoires/loa/visions/` and `.loa/grimoires/loa/lore/patterns.yaml`. They're written to be promotable into the canonical registry if/when this proposal is accepted.

### Ask

Looking for feedback on:

1. **Naming**: is `Marasa-Trois` too project-specific to `loa-vps-setup`'s Vodou-lineage? An alternative like `empirical-falsification-gate` is more generic but loses the mnemonic. The framework already has Vodou-adjacent names (`Hounfour`, `Bridgebuilder`-as-loa-agent, etc.) — happy to keep Marasa-Trois, rename if the maintainers prefer.
2. **Integration shape**: does Phase 2.6 of `reviewing-code` make sense, or would this fit better as a separate skill (`falsifying-dissent`) that the review skill optionally invokes?
3. **Scope of v1**: predicates as freeform bash strings (current reference impl) vs. constrained DSL vs. existing-test-function calls.
4. **Composition with #807 + #810**: is there an integration cycle that should bundle all three? Or land them as separate concerns?

Happy to open a PR against the framework if the maintainers want this as a feature. Filing as an issue first because the design questions above benefit from input before code.

— Filed by jani / 0xhoneyjar.xyz, with reference impl at `0xHoneyJar/loa-vps-setup` PR #3
